lib/Platform/Intl/PlatformIntlApple.mm

```diff
@@ -4,20 +4,79 @@
  * This source code is licensed under the MIT license found in the
  * LICENSE file in the root directory of this source tree.
  */
-
 #include "hermes/Platform/Intl/PlatformIntl.h"
 
 #import <Foundation/Foundation.h>
 
+NSString *u16StringToNSString(std::u16string src) {
+  auto size = src.size();
+  auto cString = (const unichar *)src.c_str();
+  return [NSString stringWithCharacters:cString length:size];
+}
+
+std::u16string nsStringToU16String(NSString *src) {
+  auto size = src.length;
+  std::u16string result;
+  result.resize(size);
+  [src getCharacters:(unichar *)&result[0] range:NSMakeRange(0, size)];
+  return result;
+}
+
 namespace hermes {
 namespace platform_intl {
 
+// Implementation of https://tc39.es/ecma402/#sec-canonicalizelocalelist
 vm::CallResult<std::vector<std::u16string>> getCanonicalLocales(
     vm::Runtime *runtime,
     const std::vector<std::u16string> &locales) {
-  return std::vector<std::u16string>{u"fr-FR", u"es-ES"};
-}
-
+  // 1. If locales is undefined, then
+  // Return a new empty List.
+  if (locales.size() == 0) {
+    return std::vector<std::u16string>();
+  }
+  // Note:: Some other major input validation occurs closer to VM in
+  // 'normalizeLocales' in JSLib/Intl.cpp
+  // 2. Let seen be a new empty List.
+  std::vector<std::u16string> seen;
+  // 3. If Type(locales) is String or Type(locales) is Object and locales has an
+  // [[InitializedLocale]] internal slot, then
+  // 4. Else
+  // We don't yet support Locale object -
+  // https://tc39.es/ecma402/#locale-objects As of now, 'locales' can only be a
+  // string list/array. 'O' is not a string array of locales
+  // 5. Let len be ? ToLength(? Get(O, "length")).
+  // 6. Let k be 0.
+  // 7. Repeat, while k < len
+  for (std::u16string locale : locales) {
+    // We don't have steps for 7a. 7b. 7c. i-iv  .. as we only allow string
+    // arrays here.. Smoke validation. Throw RangeError if input locale string
+    // is (1) empty (2) non-ASCII string.
+    if (locale.empty()) {
+      return runtime->raiseRangeError("Incorrect locale information provided");
+    }
+    if (std::find(locale.begin(), locale.end(), '_') != locale.end()) {
+      return runtime->raiseRangeError("Incorrect locale information provided");
+    }
+    // 7.c.v & 7.c.vi
+    auto tempLocale = u16StringToNSString(locale);
```

Comment:
Let's give this a more descriptive name instead of tempLocale. Perhaps something like `localeNSString`. Since this is a pointer, we should also follow the convention of using `auto *`